pipeline_dp/pipeline_operations.py

```diff
@@ -349,3 +346,213 @@ def count_per_element(self, col, stage_name: typing.Optional[str] = None):
 
     def reduce_accumulators_per_key(self, col, stage_name: str = None):
         return self.map_values(self.group_by_key(col), accumulator.merge)
+
+
+class Sentinel(Enum):
+    EOI = "EndOfIterations"
+
+
+def _multiproc_iter_input(iter_obj):
+    yield from iter_obj
+    yield Sentinel.EOI
+
+
+def _multiproc_iter_output(iter_obj):
+    for item in iter_obj:
+        if item is Sentinel.EOI:
+            break
+        yield item
+
+
+# workaround for passing lambda functions to multiprocessing
+# according to https://medium.com/@yasufumy/python-multiprocessing-c6d54107dd55
+_pool_current_func = None
+
+
+def _pool_worker_init(func):
+    global _pool_current_func
+    _pool_current_func = func
+
+
+def _pool_worker(row):
+    if row is Sentinel.EOI:
+        return Sentinel.EOI
+    return _pool_current_func(row)
+
+
+class _LazyMultiProcIterator:
+    def __init__(
+        self,
+        job: typing.Callable,
+        job_inputs: typing.Iterable,
+        chunksize: int = 1,
+        n_jobs: typing.Optional[int] = None,
+        **pool_kwargs
+    ):
+        self.job = job
+        self.chunksize = chunksize
+        self.job_inputs = job_inputs
+        self.n_jobs = n_jobs
+        self.pool_kwargs = pool_kwargs
+        self._outputs = None  # type: typing.Optional[typing.Iterator]
+        self._pool = None
+
+    def _init_pool(self):
+        self._pool = mp.Pool(
+            self.n_jobs,
+            initializer=_pool_worker_init,
+            initargs=(self.job,),
+            **self.pool_kwargs
+        )
+        return self._pool
+
+    def _trigger_iterations(self):
+        if self._outputs is None:
+            self._outputs = self._init_pool().map(
+                _pool_worker, _multiproc_iter_input(self.job_inputs), self.chunksize
+            )
+
+    def __iter__(self):
+        if isinstance(self.job_inputs, _LazyMultiProcIterator):
+            self.job_inputs._trigger_iterations()
+        self._trigger_iterations()
+        yield from _multiproc_iter_output(self._outputs)
+
+
+class _LazyMultiProcGroupByIterator(_LazyMultiProcIterator):
+    def __init__(self, job_inputs: typing.Iterable,
```

Comment:
Could you please add documentation for this class with a short explanation how it works? With mentionening that results_dict is used for collecting results between processes and that it protects us from any parallel issues.

Reply:
Done, hope to satisfaction.

Reply:
Thanks!